File: mobile-app/lib/models/analytics.dart
import 'condition_log.dart';
import 'task.dart';
import 'time_log.dart';

// 分析ロジック（純粋関数）。
// 定義は Web 版（web/src/domain/analytics.js）と一致させること。

// ---- 着手リードタイム -------------------------------------------------------

class LeadTimeSummary {
  const LeadTimeSummary({required this.sizeLabel, required this.averageDays});

  final String sizeLabel;
  final double averageDays;
}

List<LeadTimeSummary> calculateLeadTimes(List<Task> tasks) {
  final summaries = <LeadTimeSummary>[];

  for (final size in ['S', 'M', 'L']) {
    final days =
        tasks
            .where((task) {
              return task.sizeLabel == size &&
                  task.deadline != null &&
                  task.startedAt != null;
            })
            .map((task) {
              // 締切から何日前に着手されたか (deadline - startedAt)
              final diff = task.deadline!.difference(task.startedAt!);
              return diff.inHours / 24.0;
            })
            .toList();

    if (days.isEmpty) {
      summaries.add(LeadTimeSummary(sizeLabel: size, averageDays: 0));
      continue;
    }

    final total = days.fold<double>(0, (sum, value) => sum + value);
    summaries.add(
      LeadTimeSummary(sizeLabel: size, averageDays: total / days.length),
    );
  }

  return summaries;
}

// ---- 共通ヘルパー -----------------------------------------------------------

bool _isAnalyzable(Task task) => !task.isTutorialTask;

int _sumDurationSeconds(String taskId, List<TimeLog> timeLogs) {
  return timeLogs
      .where((log) => log.taskId == taskId)
      .fold<int>(0, (sum, log) => sum + log.durationSeconds);
}

// ---- 1. 見積もり精度：SML × 実働時間 ----------------------------------------

class EstimationSummary {
  const EstimationSummary({
    required this.sizeLabel,
    required this.count,
    required this.avgMinutes,
    required this.totalMinutes,
  });

  final String sizeLabel;
  final int count;
  final double avgMinutes;
  final double totalMinutes;
}

/// S/M/L ごとに、実際に費やした作業時間（timeLogs 合計）の平均を算出する。
List<EstimationSummary> calculateEstimationAccuracy(
  List<Task> tasks,
  List<TimeLog> timeLogs,
) {
  final summaries = <EstimationSummary>[];

  for (final size in ['S', 'M', 'L']) {
    final minutesList =
        tasks
            .where((task) => _isAnalyzable(task) && task.sizeLabel == size)
            .map((task) => _sumDurationSeconds(task.id, timeLogs) / 60.0)
            .where((minutes) => minutes > 0) // 実働ログがあるタスクのみ
            .toList();

    if (minutesList.isEmpty) {
      summaries.add(
        EstimationSummary(
          sizeLabel: size,
          count: 0,
          avgMinutes: 0,
          totalMinutes: 0,
        ),
      );
      continue;
    }

    final totalMinutes = minutesList.fold<double>(0, (s, m) => s + m);
    summaries.add(
      EstimationSummary(
        sizeLabel: size,
        count: minutesList.length,
        avgMinutes: totalMinutes / minutesList.length,
        totalMinutes: totalMinutes,
      ),
    );
  }

  return summaries;
}

/// S→M→L の平均作業時間が単調増加になっているか（見積もりの一貫性）を判定する。
bool isEstimationConsistent(List<EstimationSummary> accuracy) {
  final valid = accuracy.where((a) => a.count > 0).toList();
  for (var i = 1; i < valid.length; i++) {
    if (valid[i].avgMinutes < valid[i - 1].avgMinutes) return false;
  }
  return valid.length >= 2;
}

// ---- 2. 一夜漬け度 ----------------------------------------------------------

const int _crammingWindowHours = 24; // 締切この時間前以降を「一夜漬け」とみなす
const double _crammingThreshold = 0.5; // この比率以上で「一夜漬けタスク」と判定

class CrammingBySize {
  const CrammingBySize({
    required this.sizeLabel,
    required this.ratio,
    required this.count,
  });

  final String sizeLabel;
  final double? ratio;
  final int count;
}

class CrammingTask {
  const CrammingTask({
    required this.taskId,
    required this.title,
    required this.sizeLabel,
    required this.ratio,
  });

  final String taskId;
  final String title;
  final String? sizeLabel;
  final double ratio;
}

class CrammingScores {
  const CrammingScores({
    required this.overallRatio,
    required this.taskCount,
    required this.crammedTaskCount,
    required this.bySize,
    required this.topTasks,
  });

  final double? overallRatio;
  final int taskCount;
  final int crammedTaskCount;
  final List<CrammingBySize> bySize;
  final List<CrammingTask> topTasks;
}

/// タスクごとに、全作業時間のうち「締切24時間前以降（超過分含む）」に行った割合を算出する。
CrammingScores calculateCrammingScores(
  List<Task> tasks,
  List<TimeLog> timeLogs,
) {
  final perTask = <CrammingTask>[];

  for (final task in tasks) {
    if (!_isAnalyzable(task)) continue;
    final deadline = task.deadline;
    if (deadline == null) continue;

    final logs = timeLogs.where((log) => log.taskId == task.id).toList();
    final totalSec = logs.fold<int>(0, (sum, log) => sum + log.durationSeconds);
    if (totalSec <= 0) continue;

    final crammedThreshold = deadline.subtract(
      const Duration(hours: _crammingWindowHours),
    );
    final crammedSec = logs.fold<int>(0, (sum, log) {
      final start = log.startTime ?? log.endTime;
      if (start != null && !start.isBefore(crammedThreshold)) {
        return sum + log.durationSeconds;
      }
      return sum;
    });

    perTask.add(
      CrammingTask(
        taskId: task.id,
        title: task.title,
        sizeLabel: task.sizeLabel,
        ratio: crammedSec / totalSec,
      ),
    );
  }

  double? averageRatio(List<CrammingTask> list) {
    if (list.isEmpty) return null;
    return list.fold<double>(0, (s, x) => s + x.ratio) / list.length;
  }

  final bySize = ['S', 'M', 'L'].map((size) {
    final list = perTask.where((t) => t.sizeLabel == size).toList();
    return CrammingBySize(
      sizeLabel: size,
      ratio: averageRatio(list),
      count: list.length,
    );
  }).toList();

  final sorted = [...perTask]..sort((a, b) => b.ratio.compareTo(a.ratio));

  return CrammingScores(
    overallRatio: averageRatio(perTask),
    taskCount: perTask.length,
    crammedTaskCount:
        perTask.where((t) => t.ratio >= _crammingThreshold).length,
    bySize: bySize,
    topTasks: sorted.take(3).toList(),
  );
}

// ---- 3. 時間帯 × コンディション ---------------------------------------------

class TimeBand {
  const TimeBand({
    required this.key,
    required this.label,
    required this.range,
    required this.startHour,
    required this.endHour,
  });

  final String key;
  final String label;
  final String range;
  final int startHour;
  final int endHour;
}

const List<TimeBand> timeBands = [
  TimeBand(key: 'midnight', label: '深夜', range: '0-6時', startHour: 0, endHour: 6),
  TimeBand(key: 'morning', label: '午前', range: '6-12時', startHour: 6, endHour: 12),
  TimeBand(
    key: 'afternoon',
    label: '午後',
    range: '12-18時',
    startHour: 12,
    endHour: 18,
  ),
  TimeBand(key: 'night', label: '夜', range: '18-24時', startHour: 18, endHour: 24),
];

String _bandKeyOf(int hour) {
  for (final b in timeBands) {
    if (hour >= b.startHour && hour < b.endHour) return b.key;
  }
  return 'night';
}

class TimeBandCondition {
  TimeBandCondition({
    required this.key,
    required this.label,
    required this.range,
    this.good = 0,
    this.fair = 0,
    this.poor = 0,
    this.total = 0,
  });

  final String key;
  final String label;
  final String range;
  double good;
  double fair;
  double poor;
  double total;
}

/// 「どの時間帯に作業したタスクが、どんなコンディションで終わったか」を作業時間（分）で集計する。
List<TimeBandCondition> calculateConditionByTimeOfDay(
  List<Task> tasks,
  List<TimeLog> timeLogs,
  List<ConditionLog> conditionLogs,
) {
  final excludedTaskIds =
      tasks.where((t) => !_isAnalyzable(t)).map((t) => t.id).toSet();

  // taskId -> 最新のコンディション
  final conditionByTask = <String, ConditionLog>{};
  for (final log in conditionLogs) {
    final prev = conditionByTask[log.taskId];
    if (prev == null) {
      conditionByTask[log.taskId] = log;
    } else {
      final prevTime =
          prev.createdAt ?? DateTime.fromMillisecondsSinceEpoch(0);
      final curTime = log.createdAt ?? DateTime.fromMillisecondsSinceEpoch(0);
      if (curTime.isAfter(prevTime)) {
        conditionByTask[log.taskId] = log;
      }
    }
  }

  final result = timeBands
      .map(
        (b) => TimeBandCondition(key: b.key, label: b.label, range: b.range),
      )
      .toList();
  final indexByKey = {
    for (var i = 0; i < result.length; i++) result[i].key: i,
  };

  for (final log in timeLogs) {
    if (excludedTaskIds.contains(log.taskId)) continue;
    final cond = conditionByTask[log.taskId];
    final start = log.startTime;
    if (cond == null || start == null) continue;

    final minutes = log.durationSeconds / 60.0;
    if (minutes <= 0) continue;

    final bucket = result[indexByKey[_bandKeyOf(start.hour)]!];
    if (cond.condition == 'good') {
      bucket.good += minutes;
    } else if (cond.condition == 'poor') {
      bucket.poor += minutes;
    } else {
      bucket.fair += minutes;
    }
    bucket.total += minutes;
  }

  return result;
}

// ---- 3'. よく作業する時間帯（コンディション比較なし） -----------------------

class TimeBandWork {
  TimeBandWork({
    required this.key,
    required this.label,
    required this.range,
    this.total = 0,
  });

  final String key;
  final String label;
  final String range;
  double total;
}

/// 「どの時間帯にどれくらい作業したか」を作業時間（分）で時間帯別に集計する。
/// コンディションとの比較を行わない、作業量だけのシンプルな集計。
List<TimeBandWork> calculateWorkTimeByTimeOfDay(
  List<Task> tasks,
  List<TimeLog> timeLogs,
) {
  final excludedTaskIds =
      tasks.where((t) => !_isAnalyzable(t)).map((t) => t.id).toSet();

  final result = timeBands
      .map((b) => TimeBandWork(key: b.key, label: b.label, range: b.range))
      .toList();
  final indexByKey = {
    for (var i = 0; i < result.length; i++) result[i].key: i,
  };

  for (final log in timeLogs) {
    if (excludedTaskIds.contains(log.taskId)) continue;
    final start = log.startTime;
    if (start == null) continue;
    final minutes = log.durationSeconds / 60.0;
    if (minutes <= 0) continue;
    result[indexByKey[_bandKeyOf(start.hour)]!].total += minutes;
  }

  return result;
}

// ---- 4. 放置タスク検出 ------------------------------------------------------

const int _stalledThresholdDays = 3; // 最終作業からこの日数以上で「放置」と判定

class StalledTask {
  const StalledTask({
    required this.taskId,
    required this.title,
    required this.sizeLabel,
    required this.lastActivityAt,
    required this.stalledDays,
    required this.isOverdue,
    required this.deadline,
  });

  final String taskId;
  final String title;
  final String? sizeLabel;
  final DateTime? lastActivityAt;
  final double stalledDays;
  final bool isOverdue;
  final DateTime? deadline;
}

/// 「とりかかり中（DOING）のまま一定日数作業していない」タスクを検出する。
List<StalledTask> detectStalledTasks(
  List<Task> tasks,
  List<TimeLog> timeLogs, {
  DateTime? now,
  int thresholdDays = _stalledThresholdDays,
}) {
  final current = now ?? DateTime.now();
  final stalled = <StalledTask>[];

  for (final task in tasks) {
    if (!_isAnalyzable(task)) continue;
    if (task.status != TaskStatus.doing) continue;

    // 最終作業日時を求める
    final logTimes =
        timeLogs
            .where((log) => log.taskId == task.id)
            .map((log) => log.endTime ?? log.startTime)
            .whereType<DateTime>()
            .toList();

    DateTime? lastActivityAt;
    if (logTimes.isNotEmpty) {
      lastActivityAt = logTimes.reduce((a, b) => a.isAfter(b) ? a : b);
    } else {
      lastActivityAt = task.startedAt ?? task.updatedAt;
    }
    if (lastActivityAt == null) continue;

    final stalledDays =
        current.difference(lastActivityAt).inHours / 24.0;
    if (stalledDays < thresholdDays) continue;

    final deadline = task.deadline;
    stalled.add(
      StalledTask(
        taskId: task.id,
        title: task.title,
        sizeLabel: task.sizeLabel,
        lastActivityAt: lastActivityAt,
        stalledDays: stalledDays,
        deadline: deadline,
        isOverdue: deadline != null && current.isAfter(deadline),
      ),
    );
  }

  stalled.sort((a, b) => b.stalledDays.compareTo(a.stalledDays));
  return stalled;
}

File: web/src/domain/analytics.js
/**
 * Analytics Domain
 * 分析タブで使う集計ロジックを純粋関数として集約します。
 * Presentation（AnalyticsPanel）からは表示に専念させ、ここで計算を完結させます。
 * 定義は Flutter 版（mobile-app/lib/models/analytics.dart）と一致させること。
 */

// ---- 共通ヘルパー -----------------------------------------------------------

const MS_PER_DAY = 1000 * 60 * 60 * 24;
const MS_PER_HOUR = 1000 * 60 * 60;

// Firestore Timestamp / Date / 文字列のいずれでも Date に正規化する
function toDate(value) {
    if (!value) return null;
    if (value.toDate) return value.toDate();
    if (value instanceof Date) return value;
    const d = new Date(value);
    return isNaN(d.getTime()) ? null : d;
}

// 分析対象から除外するタスク（チュートリアル用ダミー）
function isAnalyzableTask(task) {
    return task && !task.isTutorialTask;
}

// 指定タスクの実働時間（秒）合計
function sumDurationSeconds(taskId, timeLogs) {
    return timeLogs
        .filter(log => log.taskId === taskId)
        .reduce((sum, log) => sum + (log.durationSeconds || 0), 0);
}

// ---- 着手リードタイム（既存仕様を移設） -------------------------------------

// S/M/L別のリードタイム算出（締切日 - 着手日の平均日数）
export function calculateLeadTimes(tasks) {
    const summaries = [];
    for (const size of ['S', 'M', 'L']) {
        const matchedTasks = tasks.filter(task =>
            task.sizeLabel === size &&
            task.deadline &&
            task.startedAt
        );

        const days = matchedTasks.map(task => {
            const deadlineDate = toDate(task.deadline);
            const startedAtDate = toDate(task.startedAt);
            if (!deadlineDate || !startedAtDate) return null;
            return (deadlineDate.getTime() - startedAtDate.getTime()) / MS_PER_DAY;
        }).filter(val => val !== null);

        if (days.length === 0) {
            summaries.push({ sizeLabel: size, averageDays: 0, count: 0 });
            continue;
        }

        const total = days.reduce((sum, val) => sum + val, 0);
        summaries.push({
            sizeLabel: size,
            averageDays: total / days.length,
            count: days.length
        });
    }
    return summaries;
}

// ---- 1. 見積もり精度：SML × 実働時間 ----------------------------------------

/**
 * S/M/L ラベルごとに、実際に費やした作業時間（timeLogs 合計）の平均を算出する。
 * 「Sのつもりが実はMより重かった」等の見積もりズレを可視化する。
 * @returns {{sizeLabel:string, count:number, avgMinutes:number, totalMinutes:number}[]}
 */
export function calculateEstimationAccuracy(tasks, timeLogs = []) {
    const summaries = [];
    for (const size of ['S', 'M', 'L']) {
        const minutesList = tasks
            .filter(task => isAnalyzableTask(task) && task.sizeLabel === size)
            .map(task => sumDurationSeconds(task.id, timeLogs) / 60)
            .filter(minutes => minutes > 0); // 実働ログがあるタスクのみ

        if (minutesList.length === 0) {
            summaries.push({ sizeLabel: size, count: 0, avgMinutes: 0, totalMinutes: 0 });
            continue;
        }

        const totalMinutes = minutesList.reduce((sum, m) => sum + m, 0);
        summaries.push({
            sizeLabel: size,
            count: minutesList.length,
            avgMinutes: totalMinutes / minutesList.length,
            totalMinutes
        });
    }
    return summaries;
}

/**
 * S→M→L の平均作業時間が単調増加になっているか（見積もりの一貫性）を判定する。
 * count>0 のサイズだけを対象に、逆転があれば false。
 */
export function isEstimationConsistent(accuracy) {
    const valid = accuracy.filter(a => a.count > 0);
    for (let i = 1; i < valid.length; i++) {
        if (valid[i].avgMinutes < valid[i - 1].avgMinutes) return false;
    }
    return valid.length >= 2;
}

// ---- 2. 一夜漬け度 ----------------------------------------------------------

const CRAMMING_WINDOW_HOURS = 24; // 締切この時間前以降の作業を「一夜漬け」とみなす
const CRAMMING_THRESHOLD = 0.5;   // この比率以上で「一夜漬けタスク」と判定

/**
 * タスクごとに、全作業時間のうち「締切24時間前以降（超過分含む）」に行った割合を算出する。
 * @returns {{
 *   overallRatio: number|null,
 *   taskCount: number,
 *   crammedTaskCount: number,
 *   bySize: {sizeLabel:string, ratio:number|null, count:number}[],
 *   topTasks: {taskId:string, title:string, sizeLabel:string|null, ratio:number}[]
 * }}
 */
export function calculateCrammingScores(tasks, timeLogs = []) {
    const perTask = [];

    for (const task of tasks) {
        if (!isAnalyzableTask(task)) continue;
        const deadline = toDate(task.deadline);
        if (!deadline) continue;

        const logs = timeLogs.filter(log => log.taskId === task.id);
        const totalSec = logs.reduce((sum, log) => sum + (log.durationSeconds || 0), 0);
        if (totalSec <= 0) continue;

        const crammedThreshold = deadline.getTime() - CRAMMING_WINDOW_HOURS * MS_PER_HOUR;
        const crammedSec = logs.reduce((sum, log) => {
            const start = toDate(log.startTime) || toDate(log.endTime);
            if (start && start.getTime() >= crammedThreshold) {
                return sum + (log.durationSeconds || 0);
            }
            return sum;
        }, 0);

        perTask.push({
            taskId: task.id,
            title: task.title,
            sizeLabel: task.sizeLabel || null,
            ratio: crammedSec / totalSec
        });
    }

    const averageRatio = (list) =>
        list.length === 0 ? null : list.reduce((s, x) => s + x.ratio, 0) / list.length;

    const bySize = ['S', 'M', 'L'].map(size => {
        const list = perTask.filter(t => t.sizeLabel === size);
        return { sizeLabel: size, ratio: averageRatio(list), count: list.length };
    });

    return {
        overallRatio: averageRatio(perTask),
        taskCount: perTask.length,
        crammedTaskCount: perTask.filter(t => t.ratio >= CRAMMING_THRESHOLD).length,
        bySize,
        topTasks: [...perTask].sort((a, b) => b.ratio - a.ratio).slice(0, 3)
    };
}

// ---- 3. 時間帯 × コンディション ---------------------------------------------

// 時間帯の定義（24時間を4分割）
export const TIME_BANDS = [
    { key: 'midnight', label: '深夜', range: '0-6時', startHour: 0, endHour: 6 },
    { key: 'morning', label: '午前', range: '6-12時', startHour: 6, endHour: 12 },
    { key: 'afternoon', label: '午後', range: '12-18時', startHour: 12, endHour: 18 },
    { key: 'night', label: '夜', range: '18-24時', startHour: 18, endHour: 24 }
];

function bandKeyOf(hour) {
    const band = TIME_BANDS.find(b => hour >= b.startHour && hour < b.endHour);
    return band ? band.key : 'night';
}

/**
 * 「どの時間帯に作業したタスクが、どんなコンディション（good/fair/poor）で終わったか」を集計する。
 * 各作業ログをタスクのコンディションに紐付け、作業時間（分）で重み付けして時間帯別に積み上げる。
 * @returns {{key:string, label:string, range:string, good:number, fair:number, poor:number, total:number}[]}
 */
export function calculateConditionByTimeOfDay(tasks, timeLogs = [], conditionLogs = []) {
    // 除外対象（チュートリアル）のタスクIDセット
    const excludedTaskIds = new Set(
        tasks.filter(t => !isAnalyzableTask(t)).map(t => t.id)
    );

    // taskId -> 最新のコンディション
    const conditionByTask = {};
    for (const log of conditionLogs) {
        const created = toDate(log.createdAt);
        const prev = conditionByTask[log.taskId];
        if (!prev || (created && prev.createdAt && created.getTime() > prev.createdAt.getTime()) || !prev.createdAt) {
            conditionByTask[log.taskId] = { condition: log.condition, createdAt: created };
        }
    }

    const result = TIME_BANDS.map(b => ({
        key: b.key, label: b.label, range: b.range, good: 0, fair: 0, poor: 0, total: 0
    }));
    const indexByKey = Object.fromEntries(result.map((r, i) => [r.key, i]));

    for (const log of timeLogs) {
        if (excludedTaskIds.has(log.taskId)) continue;
        const cond = conditionByTask[log.taskId];
        if (!cond) continue;
        const start = toDate(log.startTime);
        if (!start) continue;

        const minutes = (log.durationSeconds || 0) / 60;
        if (minutes <= 0) continue;

        const bucket = result[indexByKey[bandKeyOf(start.getHours())]];
        if (cond.condition === 'good') bucket.good += minutes;
        else if (cond.condition === 'poor') bucket.poor += minutes;
        else bucket.fair += minutes;
        bucket.total += minutes;
    }

    return result;
}

// ---- 4. 放置タスク検出 ------------------------------------------------------

const STALLED_THRESHOLD_DAYS = 3; // 最終作業からこの日数以上で「放置」と判定

/**
 * 「とりかかり中（DOING）のまま一定日数作業していない」タスクを検出する。
 * 最終作業日時 = そのタスクの timeLogs の最新 endTime（無ければ startedAt/updatedAt）。
 * @returns {{taskId:string, title:string, sizeLabel:string|null, lastActivityAt:Date|null,
 *            stalledDays:number, isOverdue:boolean, deadline:Date|null}[]}
 */
export function detectStalledTasks(tasks, timeLogs = [], options = {}) {
    const now = options.now ? toDate(options.now) : new Date();
    const thresholdDays = options.thresholdDays ?? STALLED_THRESHOLD_DAYS;
    const stalled = [];

    for (const task of tasks) {
        if (!isAnalyzableTask(task)) continue;
        if (task.status !== 'DOING') continue;

        // 最終作業日時を求める
        const logTimes = timeLogs
            .filter(log => log.taskId === task.id)
            .map(log => toDate(log.endTime) || toDate(log.startTime))
            .filter(Boolean);

        let lastActivityAt = logTimes.length > 0
            ? new Date(Math.max(...logTimes.map(d => d.getTime())))
            : (toDate(task.startedAt) || toDate(task.updatedAt));

        if (!lastActivityAt) continue;

        const stalledDays = (now.getTime() - lastActivityAt.getTime()) / MS_PER_DAY;
        if (stalledDays < thresholdDays) continue;

        const deadline = toDate(task.deadline);
        stalled.push({
            taskId: task.id,
            title: task.title,
            sizeLabel: task.sizeLabel || null,
            lastActivityAt,
            stalledDays,
            deadline,
            isOverdue: !!deadline && now.getTime() > deadline.getTime()
        });
    }

    return stalled.sort((a, b) => b.stalledDays - a.stalledDays);
}
